### ISSUE TITLE
require_relative doesn't work inside Jar with 9.3.4.0

### ISSUE THREAD

**Issue by dgolombek** [jruby/jruby#7211]
**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
jruby 9.3.4.0 (2.6.8) 2022-03-23 eff48c1ebf OpenJDK 64-Bit Server VM 25.232-b09 on 1.8.0_232-b09 +jit [x86_64-darwin]
* Operating system and platform (e.g. `uname -a`)
Darwin ip-192-168-1-155.us-west-2.compute.internal 18.7.0 Darwin Kernel Version 18.7.0: Tue Jan 12 22:04:47 PST 2021; root:xnu-4903.278.56~1/RELEASE_X86_64 x86_64
(But also reproduces in Docker with Ubuntu 18.04)

Other relevant info you may wish to add:
See [https://github.com/dgolombek/loadpath_bug](https://github.com/dgolombek/loadpath_bug) for code to reproduce

**Expected Behavior**
When compiled with JRuby 9.2.9.0 embedded, this works correctly. I see

```
["uri:classloader://loadpath_bug/lib", "/Users/david.golombek/.rvm/gems/ruby-2.6.3@global/gems/did_you_mean-1.3.0/lib", "uri:classloader:/gems/bundler-2.3.13/lib", "uri:classloader:/META-INF/jruby.home/lib/ruby/2.5/site_ruby", "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib"]
In loadpath-bug.rb before require
In loadpath-bug.rb after require
In loadpath-bug.rb before require_relative
In loadpath-bug.rb after require_relative
Starting!
```

**Actual Behavior**

With 9.3.4.0, I see

```
["uri:classloader:/loadpath_bug/lib", "/Users/david.golombek/.rvm/rubies/ruby-2.6.3/lib/ruby/gems/2.6.0/gems/did_you_mean-1.3.0/lib", "uri:classloader:/gems/bundler-2.3.13/lib", "uri:classloader:/META-INF/jruby.home/lib/ruby/2.6/site_ruby", "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib"]
In loadpath-bug.rb before require
In loadpath-bug.rb after require
In loadpath-bug.rb before require_relative
ERROR: org.jruby.embed.EvalFailedException: (ENOENT) No such file or directory - /private/tmp/lib/loadpath/bug.rb
```

As you can see, the require works, but require_relative doesn't. And it doesn't seem to be related to both bug.rb and bug/ existing -- this fails when name and directory aren't the same.

**Comment by dgolombek** [comment]
It looks like this is a more general issue with relative paths inside a jar with uri:classloader -- I ran into another piece where

`File.expand_path(File.dirname(__FILE__) + '/../../../config/ca-bundle.crt')`

was returning a relative path, not resolving this down to the root of the jar and creating an absolute path from there -- that relative path was then applied to the location where the code was executed. I can provide more details if that would be helpful.

**Comment by headius** [comment]
Could you narrow down exactly which release broke this? I'm not sure what would have been in 9.3.4.0 specifically that could cause this problem.

If you can bisect to a specific commit, that's even better!

**Comment by rsim** [comment]
We detected the same problem in our warble-packaged jar file that `File.dirname(__FILE__)` was returning the relative path for compiled files in the jar file without the `uri:classloader:/` prefix.

I tested that this loadpath_bug sample application was working in JRuby 9.2.21.0 but was failing as described here starting from JRuby 9.3.0.0.

I added additional debugging of __FILE__ before `require_relative 'bug/version'`. This is what I got with JRuby 9.2.21.0:
```
Dir.pwd = /Users/raymond/work/loadpath_bug/tmp
__FILE__ = lib/loadpath/bug.rb
File.dirname(__FILE__) = uri:classloader:/loadpath_bug/lib/loadpath
File.expand_path(File.dirname(__FILE__) + '/bug/version.rb') = uri:classloader:/loadpath_bug/lib/loadpath/bug/version.rb
```
and this is with JRuby 9.3.0.0
```
Dir.pwd = /Users/raymond/work/loadpath_bug/tmp
__FILE__ = lib/loadpath/bug.rb
File.dirname(__FILE__) = lib/loadpath
File.expand_path(File.dirname(__FILE__) + '/bug/version.rb') = /Users/raymond/work/loadpath_bug/tmp/lib/loadpath/bug/version.rb
```

So the difference is that JRuby 9.2.x is returning `File.dirname(File.dirname(__FILE__)` with the prefix `uri:classloader:/loadpath_bug/` but in JRuby 9.3.x it is missing.

I found this PR which was included in 9.3.0.0 and I'm wondering if this could be causing this difference https://github.com/jruby/jruby/pull/6239/files

And I'm wondering that probably in JRuby 9.3 it is expected that __FILE__ already should have that `uri:classloader:/` prefix.

@headius Please let me know if you can spot what is the root cause for this issue or if I should do further investigation.

**Comment by rsim** [comment]
I am now debugging `File.dirname(__FILE__)` on JRuby 9.2.21.0 and 9.3.0.0. It calls in `RubyFile.java`
```java
@JRubyMethod(required = 1, meta = true)
public static IRubyObject dirname(ThreadContext context, IRubyObject recv, IRubyObject arg) {
    Ruby runtime = context.runtime;
    RubyString filename = StringSupport.checkEmbeddedNulls(runtime, get_path(context, arg));
    String jfilename = filename.asJavaString();
```

* On JRuby 9.2.21.0 `jfilename` is `"uri:classloader:/loadpath_bug/lib/loadpath/bug.class"`
* On JRuby 9.3.0.0 `jfilename` is `"lib/loadpath/bug.rb"`

If you inspect `__FILE__` in Ruby then in both cases it is `"lib/loadpath/bug.rb"`.

So there is some change in JRuby 9.3.0.0 that removes the prefix and replaces the extension of `__FILE__` for compiled Ruby files.

**Comment by headius** [comment]
@dgolombek Are you seeing the issue with normal uncompiled .rb files?

@rsim Thanks for the investigation. I would guess that the AOT compiler is compiling in the simple path rather than compiling in logic to get the actual filename at load time. I will look into this.

**Comment by headius** [comment]
Er, scratch that... you said __FILE__ matches in both cases. so the IR code is fine. It is the `get_path` logic that seems to be stripping the URI stuff.

**Comment by rsim** [comment]
I just tested that with uncompiled .rb files I see
```
__FILE__ = uri:classloader:/loadpath_bug/lib/loadpath/bug.rb
File.dirname(__FILE__) = uri:classloader:/loadpath_bug/lib/loadpath
```
in both JRuby 9.2.21.0 and 9.3.0.0.

I will try to bisect the commit in 9.3.0.0 which is causing the difference in compiled Ruby files.

**Comment by headius** [comment]
Only affecting compiled files is...peculiar! Let me know what you find.

**Comment by rsim** [comment]
I found with git bisect that this commit https://github.com/jruby/jruby/commit/f0821fe03c5edddef18edcf8ee0190572a65ccb5 introduced this change in compiled files from
```
__FILE__ = lib/loadpath/bug.rb
File.dirname(__FILE__) = uri:classloader:/loadpath_bug/lib/loadpath
```
to
```
__FILE__ = lib/loadpath/bug.rb
File.dirname(__FILE__) = lib/loadpath
```

**Comment by headius** [comment]
I believe I have fixed all fixable paths in #7292. The only path I cannot fix is the `main` generated in precompiled scripts, since there's no way to get the path to the class from a standard Java `main(String[] args)`.

Two fixes were required:

* AOT classes produced by jrubyc (serialized IR contained in a .class file for obfuscation) needed to pass along the filename provided at **load time** into the logic to set up root scopes. The commit mentioned above hard-coded this filename into the serialized IR, preventing it from honoring path changes.
* Loading AOT-compiled bytecode needed to use the filename provided at **load time** to set the script's filename and use that to instantiate the root scopes. This is not commonly used, since AOT compilation with `jrubyc` just serializes the JRuby IR into a .class file (obfuscation only, no pregenerated bytecode).

The examples below only exercise the AOT class file from jrubyc, but I will test the bytecode AOT paths as well:

```
[] ~/projects/jruby93 $ cat blah.rb
p __FILE__

[] ~/projects/jruby93 $ jruby blah.rb
"blah.rb"

[] ~/projects/jruby93 $ jrubyc blah.rb

[] ~/projects/jruby93 $ jruby blah.class
"blah.class"

[] ~/projects/jruby93 $ mv blah.class tmp

[] ~/projects/jruby93 $ jruby tmp/blah.class
"tmp/blah.class"

[] ~/projects/jruby93 $ jruby -I . -Xaot.loadClasses -e 'require "tmp/blah"'
"/home/headius/projects/jruby93/tmp/blah.class"

[] ~/projects/jruby93 $ jruby -Xaot.loadClasses -e 'require "uri:classloader:/tmp/blah"'
"uri:classloader:/tmp/blah.class"
```
